tests/x509/limbo/test_limbo.py

```diff
@@ -0,0 +1,125 @@
+    "pedantic-rfc5280",
+    # In rare circumstances, CABF relaxes RFC 5280's prescriptions in
+    # incompatible ways. Our validator always tries (by default) to comply
+    # closer to CABF, so we skip these.
+    "rfc5280-incompatible-with-webpki",
+}
+
+LIMBO_XFAIL_TESTCASES = {
+    # We currently allow intermediate CAs that don't have AKIs, which
+    # is technically forbidden under CABF. This is consistent with what
+    # Go's crypto/x509 and Rust's webpki crate do.
+    "rfc5280::intermediate-missing-aki",
+    # We allow root CAs where the AKI and SKI mismatch, which is technically
+    # forbidden under CABF. This is consistent with what
+    # Go's crypto/x509 and Rust's webpki crate do.
+    "webpki::aki::root-with-aki-ski-mismatch",
+}
+
+
+def _get_limbo_peer(expected_peer):
+    assert expected_peer is not None
+
+    kind = expected_peer["kind"]
+    assert kind in ("DNS", "IP")
+    value = expected_peer["value"]
+    if kind == "DNS":
+        return x509.DNSName(value)
+    else:
+        return x509.IPAddress(ipaddress.ip_address(value))
+
+
+def _limbo_testcase(testcase):
+    if testcase["id"] in LIMBO_XFAIL_TESTCASES:
+        return
+
+    features = testcase["features"]
+    if features is not None and LIMBO_UNSUPPORTED_FEATURES.intersection(
+        features
+    ):
+        return
+    assert testcase["validation_kind"] == "SERVER"
+    assert testcase["signature_algorithms"] is None
+    assert testcase["extended_key_usage"] is None or testcase[
+        "extended_key_usage"
+    ] == ["serverAuth"]
+    assert testcase["expected_peer_names"] is None
+
+    trusted_certs = [
+        load_pem_x509_certificate(cert.encode())
+        for cert in testcase["trusted_certs"]
+    ]
+    untrusted_intermediates = [
+        load_pem_x509_certificate(cert.encode())
+        for cert in testcase["untrusted_intermediates"]
+    ]
+    peer_certificate = load_pem_x509_certificate(
+        testcase["peer_certificate"].encode()
+    )
+    peer_name = _get_limbo_peer(testcase["expected_peer_name"])
+    validation_time = testcase["validation_time"]
+    validation_time = (
+        datetime.datetime.fromisoformat(validation_time)
+        if validation_time is not None
+        else None
+    )
+    max_chain_depth = testcase["max_chain_depth"]
+    should_pass = testcase["expected_result"] == "SUCCESS"
+
+    verifier = PolicyBuilder(
+        time=validation_time,
+        store=Store(trusted_certs),
+        max_chain_depth=max_chain_depth,
+    ).build_server_verifier(peer_name)
+
+    try:
+        built_chain = verifier.verify(
+            peer_certificate, untrusted_intermediates
+        )
+        assert should_pass, testcase["id"]
+
+        # Assert that the verifier returns chains in [EE, ..., TA] order.
+        assert built_chain[0] == peer_certificate
+        assert built_chain[-1] in trusted_certs
+    except VerificationError:
+        assert not should_pass, testcase["id"]
```

Comment:
Rewrite so that it's `if should_pass: call verify(); else: with pytest.raises(...) call verify()`

Reply:
9ce06d0c8